components/dashboard/src/projects/ProjectVariables.tsx

```diff
@@ -0,0 +1,126 @@
+                        <ItemField>Name</ItemField>
+                        <ItemField>Value</ItemField>
+                        <ItemField>In Prebuilds?</ItemField>
+                        <ItemField>In Workspaces?</ItemField>
+                        <ItemField></ItemField>
+                    </Item>
+                    {envVars.map(variable => {
+                        return <Item className="grid grid-cols-5 items-center">
+                            <ItemField>{variable.name}</ItemField>
+                            <ItemField>****</ItemField>
+                            <ItemField>Visible</ItemField>
+                            <ItemField>{variable.censored ? 'Censored' : 'Visible'}</ItemField>
+                            <ItemField className="text-right"><button onClick={() => deleteEnvVar(variable.id)}>x</button></ItemField>
+                        </Item>
+                    })}
+                </ItemsList>
+            </>
+        }
+    </ProjectSettingsPage>;
+}
+
+function AddVariableModal(props: { project?: Project, onClose: () => void }) {
+    const [ name, setName ] = useState<string>("");
+    const [ value, setValue ] = useState<string>("");
+    const [ censored, setCensored ] = useState<boolean>(false);
+    const [ error, setError ] = useState<Error | undefined>();
+
+    const addVariable = async () => {
+        if (!props.project) {
+            return;
+        }
+        try {
+            await getGitpodService().server.setProjectEnvironmentVariable(props.project.id, name, value, censored);
+            props.onClose();
+        } catch (err) {
+            setError(err);
+        }
+    }
+
+    return <Modal visible={true} onClose={props.onClose} onEnter={() => { addVariable(); return false; }}>
+        <h3 className="mb-4">Add Variable</h3>
+        <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 py-4 flex flex-col">
+            {error && <div className="bg-gitpod-kumquat-light rounded-md p-3 text-gitpod-red text-sm mb-2">
+                {error}
+            </div>}
+            <div>
+                <h4>Name</h4>
+                <input autoFocus className="w-full" type="text" name="name" value={name} onChange={e => setName(e.target.value)} />
+            </div>
+            <div className="mt-4">
+                <h4>Value</h4>
+                <input className="w-full" type="text" name="value" value={value} onChange={e => setValue(e.target.value)} />
+            </div>
+            <div className="mt-4">
+                <CheckBox title="Visible in Prebuilds?" desc="All Project Variables are visible in Prebuilds" checked={true} disabled={true} />
+            </div>
+            <div className="mt-4">
+                <CheckBox title="Visible in Workspaces?" desc="Choose whether this Variable should be visible in Workspaces" checked={!censored} onChange={() => setCensored(!censored)} />
```

Comment:
**suggestion**: I'm probably wrong but instead of having a disabled checbox, could we reverse the defaults here and include a warning alert at the top of the modal? Does this make sense? Happy to sync[[1](https://calendly.com/gtsiolis/coworking)] to discuss more about this in depth.

| BEFORE | AFTER |
|-|-|
| <img width="1440" alt="Screenshot 2022-01-11 at 12 22 24 AM (2)" src="https://user-images.githubusercontent.com/120486/148848585-e6e85f22-514d-44ba-9f60-1148871134dd.png"> | <img width="1440" alt="Screenshot 2022-01-11 at 12 21 34 AM (2)" src="https://user-images.githubusercontent.com/120486/148848588-dddd6c66-8e7f-4a48-a0c6-61e8055d8574.png"> |

Reply:
Thanks for the visual concept! It made me change my mind about the "fake" checkbox, as I now feel the "AFTER" design could work.

- The Alert text might need a few iterations before we can convey the full story clearly / concisely / unambiguously 😋
- I like "Mask" (better than "Censor" at least) and I also like "Hide"

Reply:
I'm in favor of the "warning", too.

Two questions come to my mind, though, when reading the new text:
 - `are always visible`: to whom? A: "all project members, and everyone with read access to your repository"
 - `mask`: What does this mean? A: "< long discussion about the trade-offs >. Security-wise: nothing (yet)."
 
I fear with this "double meaning" w.r.t. visibility/masking we're facing users with yet another concept - which is **not sound**: The whole idea of ["team members are not able to see environment variables"](https://github.com/gitpod-io/gitpod/pull/7295#discussion_r783096654) and "masking" environment variables breaks due to:
 - prebuild logs (not filtered atm)
 - all tool that access your env vars during prebuild that store them somewhere in the (persisted) file system
 - prebuilds being shared between all Gitpod users (atm)

IMO we all agree that this is a hard and delicate topic. In the spirit of moving forward with this I propose we split this into:
 - introduce Project env vars with only one level of visibility (all or nothing, present or not), allow editing in the UI (because you have access to it anyway = same level of trust). All while being _very clear_ of the impact with the warning introduced above. Including that all random viewers of the project might get a workspace based on one of your prebuilds - incl. all keys in there. :upside_down_face: 
 - put the discussion about "private env vars" on hold: IMO the key thing we'd need to do here is to allow for prebuilds to have the same visibility level like the env var: project-/team-only.
 
I understand - and recall your arguments @jankeromnes - that this would reduce the usefulness of this features (to private repos, or "keys" that you do not care about being shared with the world, basically). Still, it's the only way forward I see atm that feels sound and safe in a way I could explain it to a user with a good feeling.

/cc @JanKoehnlein

Reply:
After dwelling a bit more about this - and re-reading my comment above - I might have overreacted a bit. :roll_eyes:  My apologies @jankeromnes  if that came across to harsh.

Maybe we compress the ramblings above to:
 - top warning: "Project specific environment variables might be (indirectly) accessible for everyone with read access to your repository"
 - warning underlines: "Secrets might leak if they 1.) are printed to the log, or 2.) persisted to the file system"
 - Option: "exclude from regular workspaces", checked by default (optional: showing a 2nd warning if unchecked? :innocent: )

Also, happy to discuss in a (quick) sync round tomorrow so we can ship this features! :muscle:

Reply:
Let's add warnings and enable users who want to use it.

Reply:
Many thanks @geropl for always providing consistently valuable feedback. 🙏 I feel like our discussions always remain constructive, and by working towards a consensus, the end result always becomes much better than what I could have envisioned initially. 💯

Thanks also for your helpful feedback on the warning text -- I'll be sure to reword it accordingly. 🎯

And I also agree on what you call the "false sense of security". We shouldn't ship a feature that "traps" users by encouraging them to do unsafe things, so it's important to get this right.

I'll make the proposed adjustments today, and then see whether the "exclude from regular workspaces" feature feels good to go as is (subject to review), or if it should be split into another PR (so that the non-controversial foundation and refactorings in this PR can be merged as soon as possible, while also giving us a few more cycles to improve the "exclude" feature/UX).